pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java

```diff
@@ -0,0 +1,151 @@
+  @VisibleForTesting
+  static final int MAX_NUM_OF_HISTORY_TO_KEEP = 5;
+  @VisibleForTesting
+  static final String MOST_RECENT_SUCCESS_RUN_TS = "mostRecentSuccessRunTs";
+  @VisibleForTesting
+  static final String MOST_RECENT_ERROR_RUN_MESSAGE = "mostRecentErrorRunMessage";
+
+  private final String _taskType;
+  private final String _tableNameWithType;
+  // the timestamp to error message map of the most recent several error runs
+  @Nonnull
+  private final TreeMap<Long, String> _mostRecentErrorRunMessage;
+  // the timestamp of the most recent several success runs
+  @Nonnull
+  private final List<Long> _mostRecentSuccessRunTS;
+  private final int _version;
+
+  private TaskGeneratorMostRecentRunInfo(String tableNameWithType, String taskType,
+      Map<Long, String> mostRecentErrorRunMessage, List<Long> mostRecentSuccessRunTS, int version) {
+    _tableNameWithType = tableNameWithType;
+    _taskType = taskType;
+    // sort and keep the most recent several error messages
+    _mostRecentErrorRunMessage = new TreeMap<>();
+    mostRecentErrorRunMessage.forEach((k, v) -> {
+      _mostRecentErrorRunMessage.put(k, v);
+      if (_mostRecentErrorRunMessage.size() > MAX_NUM_OF_HISTORY_TO_KEEP) {
+        _mostRecentErrorRunMessage.remove(_mostRecentErrorRunMessage.firstKey());
+      }
+    });
+    // sort and keep the most recent several timestamp
+    Queue<Long> sortedTs = new PriorityQueue<>();
+    mostRecentSuccessRunTS.forEach(e -> {
+      sortedTs.offer(e);
+      if (sortedTs.size() > MAX_NUM_OF_HISTORY_TO_KEEP) {
+        sortedTs.poll();
+      }
+    });
+    _mostRecentSuccessRunTS = new ArrayList<>();
+    while (!sortedTs.isEmpty()) {
+      _mostRecentSuccessRunTS.add(sortedTs.poll());
+    }
+    _version = version;
+  }
+
+  /**
+   * Returns the table name with type
+   */
+  public String getTableNameWithType() {
+    return _tableNameWithType;
+  }
+
+  @Override
+  public String getTaskType() {
+    return _taskType;
+  }
+
+  /**
+   * Gets the timestamp to error message map of the most recent several error runs
+   */
+  public TreeMap<Long, String> getMostRecentErrorRunMessage() {
+    return _mostRecentErrorRunMessage;
+  }
+
+  /**
+   * Adds an error run message
+   * @param ts A timestamp
+   * @param message An error message.
+   */
+  public void addErrorRunMessage(long ts, String message) {
+    _mostRecentErrorRunMessage.put(ts, message);
+    if (_mostRecentErrorRunMessage.size() > MAX_NUM_OF_HISTORY_TO_KEEP) {
+      _mostRecentErrorRunMessage.remove(_mostRecentErrorRunMessage.firstKey());
+    }
+  }
+
+  /**
+   * Gets the timestamp of the most recent several success runs
+   */
+  public List<Long> getMostRecentSuccessRunTS() {
+    return _mostRecentSuccessRunTS;
+  }
+
+  /**
+   * Adds a success task generating run timestamp
+   * @param ts A timestamp
+   */
+  public void addSuccessRunTs(long ts) {
+    _mostRecentSuccessRunTS.add(ts);
```

Comment:
Q: any need to keep _mostRecentSuccessRunTS sorted?

Reply:
Simplifies the eviction is all. Also the APi response is easier to work with (the most recent timestamp is the one of the top)

Reply:
Does this indicate a successful run (completion) of the task, or a successful generation of the task? If the task fails in minion, does your API address it?

Reply:
This only indicates the successful generation of a the task. This API is exclusively for task generation